data/virt_autotest/guest_console_monitor.sh

```diff
@@ -0,0 +1,247 @@
+fi
+
+##CONFIG PART
+START_MONITOR="0"
+END_MONITOR="0"
+MONITOR_LOCKED_FILE="/tmp/GUEST_CONSOLE_COLLECTOR_SEMAPHORE"
+timestamp="`date '+%Y%m%d%H%M%S'`"
+###CHECK EXISTED GUEST
+vm_guestnames_types="sles"
+get_vm_guestnames_inactive=`virsh list --inactive | grep "${vm_guestnames_types}" | awk '{print $2}'`
+vm_guestnames_inactive_array=$(echo -e ${get_vm_guestnames_inactive})
+get_vm_guestnames=`virsh list  --all | grep "${vm_guestnames_types}" | awk '{print $2}'`
+vm_guestnames_array=$(echo -e ${get_vm_guestnames})
+vmguest=""
+vmguest_failed="0"
+
+#Parse input arguments. -s or -e must have values
+while getopts "seh" OPTION; do
+  case "${OPTION}" in
+    s)
+      START_MONITOR="1"
+      ;;
+    e)
+      END_MONITOR="1"
+      ;;
+    h)
+      help_usage | tee -a ${monitor_log_file}
+      exit 1
+      ;;
+    *)
+      help_usage | tee -a ${monitor_log_file}
+      exit 1
+      ;;
+  esac
+done
+
+#FUNCTION PART
+function get_console() {
+
+local vmguest=$1
+
+expect -c "
+set timeout 60
+
+##hide echo
+log_user 0
+spawn -noecho virsh console ${vmguest}
+
+#wait connection
+sleep 3
+send \"\r\n\r\n\r\n\"
+
+#condition expect
+expect {
+        \"*login:\" {
+                send \"root\r\"
+                exp_continue
+        }
+        -nocase \"password:\" {
+                send \"novell\r\"
+                exp_continue
+        }
+        \"*:~ #\" {
+                send -- \"ip route get 1\r\"
+        }
+        timeout {
+                send -- \"exit\r\"
+                exp_continue        
+        }                 
+}
+
+## -1 means never timeout
+set timeout -1
+
+expect -re {dev.*\s([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})[^0-9]}
+
+if {![info exists expect_out(1,string)]} {
+        puts \"Match did not happen :(\"
+}
+set output \$expect_out(buffer)
+expect eof
+
+puts \"\$output\"
+"
+
+}
+
+function open_console() {
+
+    #open guest console
+    for vmguest in ${vm_guestnames_state_running_array[@]};do
+        get_console ${vmguest} >> ${LOG_DIR}/${vmguest}_console_${timestamp} 2>&1 &
+    done 
+}
+
+function close_console() {
+
+    #close guest console
+    echo -e "Kill PIDs $(basename $0)\n" >> ${monitor_log_file}
+    pid=`ps aux | grep "$(basename $0) -s" | grep -v "grep"| awk '{print $2}'`
+    if [[ -n $pid ]];then
+        kill -9 $pid >> /dev/null 2>&1;sync
+    else
+        echo -e "PIDs $(basename $0) were killed\n" >> ${monitor_log_file}
+    fi
+
+    echo -e "Kill all associated "virsh console" processes\n" >> ${monitor_log_file}
+
+    while :
+    do
+        #wait for checking all associated "virsh console" processes"
+        sleep 3
+
+        #Check all associated "virsh console" processes
+        ps aux | grep "virsh console" | grep -v "grep" >> /dev/null 2>&1
+        if [[ $? -eq 0 ]];then 
+            sub_pid=`ps aux | grep "expect -c" | grep -v "grep"| awk '{print $2}'`
+            if [[ -n $sub_pid ]];then
+                kill -9 $sub_pid >> /dev/null 2>&1;sync
```

Comment:
From your description, you give below chart. From it, we can see that for a single guest, there are 1 virsh console process , and 2 expect processes as well as other virsh processes, which you can tell from different process ids. So in current cleanup, you actually lack the killing of some processes which will render orphan processes.

I did an experiment locally. There can be an easier way by "kill -9 -pgid". This command will kill all processes which have the same process group id. You can have a check whether all the processes created by "guest_console_monitor.sh -s" have the same pgid which should be just the process id of the ancestor process "guest_console_monitor.sh -s". You can do it by following cmd:

ps -ef |grep "guest_console_monitor.sh -s" //find its process, let's call it [monitor_process_id]
pstree -pal  [monitor_process_id] //find its process tree(all child processes)
ps -eo pid,ppid,pgid,comm|grep -E "all child processes id separating with |" // check the pgid column, whether it is  [monitor_process_id] for all. IF YES, then use my suggested way directly can simplify your kill process by doing only "kill -9 -[monitor_process_id]".


```

# pstree -p
...
           ├─guest_console_m(3542)─┬─guest_console_m(3750)───expect(3754)─┬─virsh(3758)───{virsh}(3+
           │                       │                                      └─{expect}(3763)
           │                       ├─guest_console_m(3751)───expect(3755)─┬─virsh(3757)───{virsh}(3+
           │                       │                                      └─{expect}(3764)

```

Reply:
Cool! will confirm this kill process by doing only "kill -9 -[monitor_process_id]" as your mentioned.

Reply:
Quick update

```
# pstree -a
...
  ├─bash -x guest_console_monitor.sh -s
  │   ├─bash -x guest_console_monitor.sh -s
  │   │   └─expect -c...
  │   │       ├─virsh console sles-15-sp2-64-fv-def-net
  │   │       │   └─{virsh}
  │   │       └─{expect}
  │   ├─bash -x guest_console_monitor.sh -s
  │   │   └─expect -c...
  │   │       ├─virsh console sles-15-sp2-64-pv-def-net
  │   │       │   └─{virsh}
  │   │       └─{expect}
  │   └─sleep 3

# ps aux | grep guest_console_monitor.sh -s| grep -v grep 
root     25132  0.3  0.1  13268  2044 pts/0    S    04:19   0:02 bash -x guest_console_monitor.sh -s - 
root     25142  0.0  0.1  13268  2012 pts/0    S    04:19   0:00 bash -x guest_console_monitor.sh -s
root     25143  0.0  0.1  13268  2012 pts/0    S    04:19   0:00 bash -x guest_console_monitor.sh -s

# pstree -p
systemd(1)─┬─agetty(1324)
...
           ├─bash(25132)─┬─bash(25142)───expect(25145)─┬─virsh(25147)───{virsh}(25155)
           │             │                             └─{expect}(25153)
           │             ├─bash(25143)───expect(25146)─┬─virsh(25149)───{virsh}(25156)
           │             │                             └─{expect}(25154)
           │             └─sleep(29253)
```

take an easier solution
`ps -eo pid,cmd | grep -E 'guest_console_monitor.sh -s|expect -c|virsh console'`

Reply:
@alice-suse , please help me make sure if there was new comments for this PR. 

if not, please help me merge this PR. thank you so much for your great help.

Reply:
Can you help explain why there are multiple "guest_console_monitor.sh -s" processes, 25132, 25142? Besides, when you kill processes during stop monitor, how can the "{expect}" process be killed? I only see you grep "expect -c", how about the process " {expect}(25154) " in your "pstree -p" output?

Reply:
> Cool! will confirm this kill process by doing only "kill -9 -[monitor_process_id]" as your mentioned.

Can you share why you abandon this method finally?

Reply:
> Can you help explain why there are multiple "guest_console_monitor.sh -s" processes, 25132, 25142
.? 
As mentioned, the other guest_console_monitor.sh -s processes, (such as 25132, 25142 ) which ran as background too, was trigger by **open_console()** 

Besides, when you kill processes during stop monitor, how can the "{expect}" process be killed? I only see you grep "expect -c", how about the process " {expect}(25154) " in your "pstree -p" output?
{expect}(25154)  was created from expect -c command (e.g get_console func) as below: 
#condition expect
```
expect {
        \"*login:\" {
                send \"root\r\"
                exp_continue
        }
        -nocase \"password:\" {
                send \"novell\r\"
                exp_continue
        }
        \"*:~ #\" {
                send -- \"ip route get 1\r\"
        }
        timeout {
                send -- \"exit\r\"
                exp_continue        
        }                 
}
```

Reply:
> > Cool! will confirm this kill process by doing only "kill -9 -[monitor_process_id]" as your mentioned.
> 
> Can you share why you abandon this method finally?

after trigger `kill -9 - [monitor_process_id]`, found that just killed only  [monitor_process_id] on vm host, the legacy child processes - ran was background pids were still keep existed. 
imho, kill this [monitor_process_id] just killed | effect [monitor_process_id]-self and sleep(29253), do not effect| kill other legacy child processes - ran were background pids.   
so, abandon this method finally.

Reply:
> > > Cool! will confirm this kill process by doing only "kill -9 -[monitor_process_id]" as your mentioned.
> > 
> > 
> > Can you share why you abandon this method finally?
> 
> imho, `kill -9 -[monitor_process_id] `just killed | effect [monitor_process_id]-self and sleep(29253), do not effect| kill other legacy child processes - ran were background pids (were created from get_console func).
> so, abandon this method finally.

Okay, I do not see your output for ps -eo pid,ppid,pgid,command|grep -E "all child processes id separating with |". But I guess it is because they do not have same pgid.

Reply:
yes, there was the other pgid:
For example:
there was 28130 as [monitor_process_id] as below:
```
# pstree -pa 28130
bash,28130 -x guest_console_monitor.sh -s
  ├─bash,28141 -x guest_console_monitor.sh -s
  │   └─expect,28143 -c...
  │       ├─virsh,28146 console sles-15-sp2-64-fv-def-net
  │       │   └─{virsh},28155
  │       └─{expect},28152
  ├─bash,28142 -x guest_console_monitor.sh -s
  │   └─expect,28145 -c...
  │       ├─virsh,28147 console sles-15-sp2-64-pv-def-net
  │       │   └─{virsh},28154
  │       └─{expect},28153
  └─sleep,28263 3
```
output for ps -eo pid,ppid,pgid,command|grep -E "all child processes id separating with |"
```
# ps -eo pid,ppid,pgid,command|grep -E "28130|28141|28143|28146|28155|28152|28142|28145|28147|28154|28153|28263"
28130     1 28103 bash -x guest_console_monitor.sh -s
28141 28130 28103 bash -x guest_console_monitor.sh -s
28142 28130 28103 bash -x guest_console_monitor.sh -s
28143 28141 28103 expect -c  set timeout 60  ##hide echo log_user 0 spawn -noecho virsh console sles-15-sp2-64-fv-def-net  #wait connection sleep 3 send "\r\n\r\n\r\n"  #condition expect expect {         "*login:" {                 send "root\r"                 exp_continue         }         -nocase "password:" {                 send "novell\r"                 exp_continue         }         "*:~ #" {                 send -- "ip route get 1\r"         }         timeout {                 send -- "exit\r"                 exp_continue                 }                  }  ## -1 means never timeout set timeout -1  expect -re {dev.*\s([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})[^0-9]}  if {![info exists expect_out(1,string)]} {         puts "Match did not happen :(" } set output $expect_out(buffer) expect eof  puts "$output" 
28145 28142 28103 expect -c  set timeout 60  ##hide echo log_user 0 spawn -noecho virsh console sles-15-sp2-64-pv-def-net  #wait connection sleep 3 send "\r\n\r\n\r\n"  #condition expect expect {         "*login:" {                 send "root\r"                 exp_continue         }         -nocase "password:" {                 send "novell\r"                 exp_continue         }         "*:~ #" {                 send -- "ip route get 1\r"         }         timeout {                 send -- "exit\r"                 exp_continue                 }                  }  ## -1 means never timeout set timeout -1  expect -re {dev.*\s([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})[^0-9]}  if {![info exists expect_out(1,string)]} {         puts "Match did not happen :(" } set output $expect_out(buffer) expect eof  puts "$output" 
28146 28143 28146 virsh console sles-15-sp2-64-fv-def-net
28147 28145 28147 virsh console sles-15-sp2-64-pv-def-net
28852 28130 28103 sleep 3
```
So, try to kill this 28130 as [monitor_process_id] , then found the legacy process were still existed:
```
  ├─bash,28141 -x guest_console_monitor.sh -s
  │   └─expect,28143 -c...
  │       ├─virsh,28146 console sles-15-sp2-64-fv-def-net
  │       │   └─{virsh},28155
  │       └─{expect},28152
  ├─bash,28142 -x guest_console_monitor.sh -s
  │   └─expect,28145 -c...
  │       ├─virsh,28147 console sles-15-sp2-64-pv-def-net
  │       │   └─{virsh},28154
  │       └─{expect},28153
```

Reply:
> > Can you help explain why there are multiple "guest_console_monitor.sh -s" processes, 25132, 25142
> > .?
> 
> **As mentioned, the other guest_console_monitor.sh -s processes, (such as 25132, 25142 ) which ran as background too, was trigger by **open_console()****
> 

Yes, you are right. Sorry I missed the background sign in the function call.

> Besides, when you kill processes during stop monitor, how can the "{expect}" process be killed? I only see you grep "expect -c", how about the process " {expect}(25154) " in your "pstree -p" output?
> 
> **{expect}(25154) was created from expect -c command (e.g get_console func) as below:**
> 
> ```
> expect {
>         \"*login:\" {
>                 send \"root\r\"
>                 exp_continue
>         }
>         -nocase \"password:\" {
>                 send \"novell\r\"
>                 exp_continue
>         }
>         \"*:~ #\" {
>                 send -- \"ip route get 1\r\"
>         }
>         timeout {
>                 send -- \"exit\r\"
>                 exp_continue        
>         }                 
> }
> ```

This explains how the process created, but it does not answer why do not you kill it separately? When you kill "expect -c" process only, can the expect process be killed(different process id from expect -c)?

Reply:
sure, had confirmed that after killed process id from expect -c command, legacy process expect command was created from `expect -c` will be killed together. 
So, just kill `expect -c `command here